sql/mdl.cc

```diff
@@ -304,6 +357,148 @@ bool Deadlock_detection_visitor::enter_node(MDL_context *node)
   return m_found_deadlock;
 }
 
+char *thd_security_context(MYSQL_THD thd, char *buffer, size_t length,
+                           size_t max_query_len) {
+  String str(buffer, length, &my_charset_latin1);
+  Security_context *sctx = &thd->main_security_ctx;
+  char header[256];
+  size_t len;
+  /*                                                                                              
+    The pointers thd->query and thd->proc_info might change since they are                        
+    being modified concurrently. This is acceptable for proc_info since its                       
+    values doesn't have to very accurate and the memory it points to is static,                   
+    but we need to attempt a snapshot on the pointer values to avoid using NULL                   
+    values. The pointer to thd->query however, doesn't point to static memory                     
+    and has to be protected by LOCK_thd_query or risk pointing to                                 
+    uninitialized memory.                                                                      
+  */
+  const char *proc_info = thd->proc_info;
+
+  len = snprintf(header, sizeof(header),
+                 "MySQL thread id %lu, OS thread handle %lu, query id %lu",
+                 (ulong)thd->thread_id, (ulong)thd->real_id, (ulong)thd->query_id);
```

Comment:
The ulong cast is not a good idea. Especially with thd->read_id, which on some OS is a struct. It is also not a good idea with the other members that are 64bit. ulong is , as we know, not always 64 bit (not even if OS is 64bit). ulong  is the least portable datatype, so maybe you avoid in places it is unnecessary? luckily, "long long" appears to be better portable, and there are %llu and %lld for it.

Reply:
Yes, there are potential overflows for different operating system platforms. We re-writed it according to the portable format.

Reply:
Not only that, THD::real_id is pthread_t, which is not required to be a number, but can be a struct, and in this case you can't cast it even to ulonglong.   https://stackoverflow.com/questions/1759794/how-to-print-pthread-t discusses exactly this problem.  I do know that we came across this problem in the past, although I can't recall the OS we had this problem (a traditional Unix,  maybe Solaris,  it might be less relevant now)

THD::os_thread_id is  more meaningful and can be output not problem, although it is only defined on Windows and Linux

Reply:
Initially I referenced the code from functions like 'MYSQL_QUERY_LOG::write' like below
    if (my_b_printf(&log_file,
                    "# Thread_id: %lu  Schema: %s  QC_hit: %s\n"
                    "# Query_time: %s  Lock_time: %s  Rows_sent: %lu  Rows_examined: %lu\n"
                    "# Rows_affected: %lu  Bytes_sent: %lu\n",
                    (ulong) thd->thread_id, thd->get_db(),
                    ((thd->query_plan_flags & QPLAN_QC) ? "Yes" : "No"),
                    query_time_buff, lock_time_buff,
                    (ulong) thd->get_sent_row_count(),
                    (ulong) thd->get_examined_row_count(),
                    (ulong) thd->get_affected_rows(),
                    (ulong) (thd->status_var.bytes_sent - thd->bytes_sent_old)))
Codes here are not bracketed with macro and maybe this writing is platform independ.
So thread any way to print it in common? Or maybe we encapsulate a function the thread_id to string print?

Reply:
Sorry, I found the my mistake. I don't found the thd_security_context function in marriadb maybe for my mind-mess. So I copied it from MySQL code. But the function thd_security_context already exists. So I think call this marriadb implement will resolve this problem.

Reply:
Ohhh, in function thd_get_error_context_description, the real_id and thread_id are also casted to ulong type.😓